{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"GATK210929.ipynb","provenance":[],"collapsed_sections":[],"authorship_tag":"ABX9TyMGNRnWyURS7tOwYm0GITY2"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","metadata":{"id":"vnlgR4s_Zq3L"},"source":["## 00-1.Mount Google Drive"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"iWegYhskQYIb","executionInfo":{"status":"ok","timestamp":1632792621955,"user_tz":-540,"elapsed":19883,"user":{"displayName":"­유준기 / 학생 / 협동과정 생물정보학전공","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"15162318757091042303"}},"outputId":"9a77295f-7675-484f-dc1b-f03f9e8a687e"},"source":["from google.colab import drive\n","drive.mount('/content/drive') "],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/drive\n"]}]},{"cell_type":"markdown","metadata":{"id":"xypLNdaSaTCk"},"source":["## 00-2.Open terminal"]},{"cell_type":"code","metadata":{"id":"JkCTuZmZaNAy"},"source":["!pip install kora\n","from kora import console\n","console.start()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"UAXc5QxgatDM"},"source":["## 01. Installation"]},{"cell_type":"markdown","metadata":{"id":"e0hZ4EkuZpOL"},"source":["# Install java, bwa, GATK, picard, samtools    "]},{"cell_type":"code","metadata":{"id":"-fELc1hsavTw"},"source":["%%bash\n","wget https://github.com/broadinstitute/gatk/releases/download/4.2.2.0/gatk-4.2.2.0.zip \n","unzip gatk-4.2.2.0\n"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"jWVQOXFoDVCA"},"source":["%%bash\n","wget https://sourceforge.net/projects/bio-bwa/files/bwa-0.7.17.tar.bz2/download\n","tar -xvf /content/drive/download/bwa-0.7.17.tar.bz2\n","\n","#move to bwa folder \n","make "],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"Kbt3q10jBOB4"},"source":["import os       #importing os to set environment variable\n","def install_java():\n","  !apt-get install -y openjdk-8-jdk-headless -qq > /dev/null      #install openjdk\n","  os.environ[\"JAVA_HOME\"] = \"/usr/lib/jvm/java-8-openjdk-amd64\"     #set environment variable\n","  !java -version       #check java version\n","install_java()"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"kpBOaeyvaU9H"},"source":["%%bash\n","wget https://sourceforge.net/projects/picard/files/picard-tools/1.73/picard-tools-1.73.zip\n","unzip picard-tools-1.73.zip\n"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"0lsy6Pr2ajAB"},"source":["%%bash\n","wget https://github.com/samtools/samtools/releases/download/1.3.1/samtools-1.3.1.tar.bz2\n","tar -xf samtools-1.3.1.tar.bz2\n","\n","#"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"GllO4QSOfE0a"},"source":["## 02. Initial Mapping (global alignment)"]},{"cell_type":"code","metadata":{"id":"sD-ii_DGhUGQ"},"source":["%%bash\n","export Drive=/content/drive/mydrive\n","bwa mem -t 8 $Drive/ucsc.hg19.fasta $Drive/illumina.fastq > illumina.sam     #-t: number of threads\n","samtools view -bhS illumina.sam > illumina_raw.bam\n","samtools sort illumina_raw.bam -o illumina_sorted.bam\n","samtools index illumina_sorted.bam\n","\n","#sam: sequence alignment/map format. bam: binary version of sam "],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"AY9CEo9jHyYy"},"source":["## 02-1. Deduplication"]},{"cell_type":"code","metadata":{"id":"1BAxluETH2C-"},"source":["%%bash\n","java -jar picard.jar MarkDuplicates \\\n","I = illumina_sorted.bam \\\n","O = illumina_dedup_sorted.bam \\\n","M = Marked_dup_metrics.txt \\\n","REMOVE_DUPLICATES = true"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Njt6CmaN8Tbz"},"source":["## 03. Base Quality Score Recalibration (BQSR)"]},{"cell_type":"code","metadata":{"id":"HOtL57Z3F04_"},"source":["%%bash\n","gatk --java-options \"-Xmx8G\" BaseRecalibrator \\\n","-I illumina_dedup_sorted.bam \\\n","-R $Drive/ucsc.hg19.fasta \\\n","--known-sites $Drive/dbsnp_138.hg19.vcf \\\n","-O illumina_recal.table \n","\n","gatk --java-options \"-Xmx8G\" ApplyBQSR \\\n","-I illumina_dedup_sorted.bam \\\n","--bqsr-recal-file illumina_recal.table\\\n","-O illumina_postrecal.bam\n","\n","\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"bK4ZqFpbJCy1"},"source":["## Variant Calling with HaplotypeCaller"]},{"cell_type":"code","metadata":{"id":"wU7FeRWRJGEX"},"source":["%%bash\n","gatk --java-options \"-Xmx8G\" HaplotypeCaller \\ \n","-I illumina_postrecal.bam \\ \n","-R $Drive/ucsc.hg19.fasta \\\n","-O illumina.vcf.gz \\ \n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"lBGNaNK2JtgW"},"source":["## Filtering SNV/INDEL"]},{"cell_type":"code","metadata":{"id":"m13vFp7YJw88"},"source":["%%bash\n","gatk –java-options ”-Xmx8G” SelectVariants \\\n","–tmp-dir=$Drive/tmp \\\n","-V illumina.vcf.gz \\ \n","-select-type SNP \\ -O SNV_illumina.vcf.gz\n","\n","gatk –java-options ”-Xmx8G” SelectVariants \\\n","–tmp-dir=$Drive/tmp \\ \n","-V illumina.vcf.gz \\\n","-select-type INDEL \\ -O INDEL_illumina.vcf.gz\n","\n","gatk –java-options ”-Xmx8G” VariantFiltration \\\n","–tmp-dir=$Drive/tmp \\ \n","-V SNV_illumina.vcf.gz\n","-filter “QD < 2.0” –filter-name “QD2” \\\n","-filter “DP < 10.0” –filter-name “LowDP10” \\ \n","-filter “QUAL < 30.0” –filter-name “QUAL30” \\\n","-filter “SOR > 3.0” –filter-name “SOR3” \\\n","-filter “FS > 60.0” –filter-name “FS60” \\\n","-filter “MQ < 40.0” –filter-name “MQ40” \\\n","-filter “MQRankSum < -12.5” –filter-name “MQRankSum-12.5” \\\n","-filter “ReadPosRankSum < -8.0” –filter-name “ReadPosRankSum-8” \\\n","-O Filtered_SNV_illumina.vcf.gz\n","\n","gatk –java-options ”-Xmx8G” VariantFiltration\n","–tmp-dir=$Drive/tmp \\\n","-V INDEL_illumina.vcf.gz \\ \n","-filter “QD < 2.0” –filter-name “QD2”\\\n","-filter “DP < 10.0” –filter-name “LowDP10” \\\n","-filter “QUAL < 30.0” –filter-name “QUAL30” \\\n","-filter “FS > 200.0” –filter-name “FS200” \\\n","-filter “ReadPosRankSum < -20.0” –filter-name “ReadPosRankSum-20” \\\n","-O Filtered_INDEL_illumina.vcf.gz"],"execution_count":null,"outputs":[]}]}